dali/operators/random/normal_distribution_cpu.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/random/rng_base_cpu.h"
+#include "dali/pipeline/util/operator_impl_utils.h"
+
+#define DALI_NORMDIST_TYPES (uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, uint64_t, \
+                             int64_t, float16, float, double)
+
+namespace dali {
+
+DALI_SCHEMA(NewNormalDistribution)
+    .DocStr(R"code(Generates random numbers following a normal distribution.
+
+The shape of the generated data can be either specified explicitly with a ``shape`` argument,
+or chosen to match the shape of the input, if provided. If none are present, a single number is
+generated.
+)code")
+    .NumInput(0, 1)
+    .NumOutput(1)
+    .AddOptionalArg("mean", R"code(Mean of the distribution.)code",
+                    0.f, true)
+    .AddOptionalArg("stddev",
+                    R"code(Standard deviation of the distribution.)code",
+                    1.f, true)
+    .AddParent("RNGAttr");
+
+
+class NormalDistributionCPU : public RNGBaseCPU<NormalDistributionCPU> {
+ public:
+  explicit NormalDistributionCPU(const OpSpec &spec)
+      : RNGBaseCPU<NormalDistributionCPU>(spec), mean_("mean", spec), stddev_("stddev", spec) {}
+
+  ~NormalDistributionCPU() override = default;
+
+  void SetupImplImpl(const OpSpec &spec, const workspace_t<CPUBackend> &ws, int nsamples) {
+    mean_.Read(spec, ws, nsamples);
+    stddev_.Read(spec, ws, nsamples);
+    dist_.clear();
+    dist_.reserve(nsamples);
+    for (int s = 0; s < nsamples; s++) {
+      dist_.emplace_back(mean_[s], stddev_[s]);
+    }
+  }
+
+  template <typename T>
+  void Generate(span<T> out, int sample_idx, std::mt19937_64& rng) {
+    auto &dist = dist_[sample_idx];
+    for (auto &value : out)
+      value = ConvertSat<T>(dist(rng));
+  }
```

Comment:
I was thinking about generating data in parallel (even for one sample) - this could be achieved by using the per-sample generator as a source of entropy for per-chunk generators. This behavior could be useful for large data samples and small batch sizes (random number generation isn't exactly cheap).
```C++
template <typename T, typename RNG, typename Dist>
void Generate(span<T> out, ThreadPool &tp, RNG &rng, Dist &&dist) {
  if (out.size() < kThreshold) {
    tp.AddWork([&rng, out, dist]() {
    // simple loop
    });
  } else {
    int chunks = out.size() / kChunkSize;
    std::array<uint32_t, kNumChunkSeeds> seed;
    for (int c = 0; c < chunks; c++) {
      auto start = out.size() * c / chunks;
      auto end = out.size() * (c + 1)  / chunks;
      auto chunk = make_span(out.begin() + start, out.begin() + end);
      for (auto &s : seed)
        s = rng();
      std::mt19937_64 chunk_rng(std::seed_seq(seed.begin(), seed.end()));  // or however seed_seq is used....
      tp.AddWork([=] {
         // loop over chunk
      });
    } 
    // simple loop
    });
  }
}
```

Reply:
done in the base class